### PR TITLE
refactor: adds `StorageReader` / `HashingWriter` / `HistoryWriter`

### PR DESCRIPTION
**Pull request by joshieDo**
Adds the following provider traits and moves the storage related methods into their `DatabaseProvider` implementations:
* `StorageReader`
* `HashingWriter` (will move the account methods in another PR as discussed )
* `HistoryWriter`(will move the account methods in another PR as discussed )

Also, removed the `TransactionError` which was tied to the previous `struct Transaction`. The tests inside `transaction.rs` shall be moved elsewhere in another PR.

### CODE REVIEW

File: crates/interfaces/src/provider.rs [hunk 1]
```diff
@@ -4,6 +4,8 @@ use reth_primitives::{Address, BlockHash, BlockHashOrNumber, BlockNumber, TxNumb
 #[allow(missing_docs)]
 #[derive(Debug, thiserror::Error, PartialEq, Eq, Clone)]
 pub enum ProviderError {
+    #[error(transparent)]
+    Database(#[from] crate::db::DatabaseError),
```
Comment (mattsse) [inline crates/interfaces/src/provider.rs:8]: I like this, 

I'd really like to use Provider error everywhere: #2719

File: crates/storage/provider/src/traits/storage.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+use auto_impl::auto_impl;
+use reth_interfaces::Result;
+use reth_primitives::{Address, BlockNumber, StorageEntry, H256};
+
+/// Storage reader
+#[auto_impl(&, Arc, Box)]
+pub trait StorageReader: Send + Sync {
+    /// Get plainstate storages from an Address.
+    fn basic_storages(
```
Comment (mattsse) [inline crates/storage/provider/src/traits/storage.rs:14]: I find this naming here a bit confusing.

what does `basic` mean here?

`basic` returns `plainstate` is not very intuitive to me

File: crates/storage/provider/src/transaction.rs [hunk 3]
```diff
@@ -3,47 +3,6 @@ use reth_primitives::{BlockHash, BlockNumber, H256};
...
 use std::fmt::Debug;
 
-/// An error that can occur when using the transaction container
-#[derive(Debug, PartialEq, Eq, Clone, thiserror::Error)]
-pub enum TransactionError {
-    /// The transaction encountered a database error.
-    #[error(transparent)]
-    Database(#[from] DbError),
-    /// The transaction encountered a database integrity error.
-    #[error(transparent)]
```
Comment (mattsse) [inline crates/storage/provider/src/transaction.rs:13]: is this now fully submerged into ProviderError?

**Review by mattsse — CHANGES_REQUESTED** [review]
lgtm overall.

nit re naming

File: crates/storage/provider/src/traits/storage.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    fn changed_storages(
+        &self,
+        range: RangeInclusive<BlockNumber>,
+    ) -> Result<BTreeMap<Address, BTreeSet<H256>>>;
+
+    /// Iterate over storage changesets and return all storage slots that were changed alongside
+    /// each specific set of blocks.
+    ///
+    /// NOTE: Get inclusive range of blocks.
+    fn changed_storage_and_blocks_with_range(
```
Comment (shekhirin) [inline crates/storage/provider/src/traits/storage.rs:29]: if `with_range` means that we're passing the range, it's the case for `changed_storages` too. I think something like that may be less confusing
```suggestion
    fn changed_storages_with_blocks(
```

File: crates/storage/provider/src/traits/storage.rs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+};
+
+use auto_impl::auto_impl;
+use reth_interfaces::Result;
+use reth_primitives::{Address, BlockNumber, StorageEntry, H256};
+
+/// Storage reader
+#[auto_impl(&, Arc, Box)]
+pub trait StorageReader: Send + Sync {
+    /// Get plainstate storages from an Address.
```
Comment (shekhirin) [inline crates/storage/provider/src/traits/storage.rs:13]: I think it should be plural and mentioning the slots, as we can retrieve multiple slots per multiple addresses
```suggestion
    /// Get plainstate storages for addresses and storage keys.
```

File: crates/storage/provider/src/transaction.rs (hunk 3 above)
Comment (joshieDo) [reply crates/storage/provider/src/transaction.rs:13]: yes

File: crates/storage/provider/src/traits/storage.rs (hunk 4 above)
Comment (joshieDo) [reply crates/storage/provider/src/traits/storage.rs]: added `with_range` to `changed_storages` instead to be similar to the `AccountExtReader`

File: crates/storage/provider/src/traits/storage.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    fn changed_storages_with_range(
+        &self,
+        range: RangeInclusive<BlockNumber>,
+    ) -> Result<BTreeMap<Address, BTreeSet<H256>>>;
+
+    /// Iterate over storage changesets and return all storage slots that were changed alongside
+    /// each specific set of blocks.
+    ///
+    /// NOTE: Get inclusive range of blocks.
+    fn changed_storages_and_blocks_with_range(
```
Comment (gakonst) [inline crates/storage/provider/src/traits/storage.rs:29]: beaut

**Review by gakonst — APPROVED** [review]

**Review by shekhirin — APPROVED** [review]

**Comment by shekhirin** [comment]
pending @mattsse approval as merge is blocked until "the requested changes are addressed" 😄

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**